File: cards.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>

#include "cards.h"

const int NUM_CARDS = 52;

void CARD_to_string(Card* c, char* buffer) {
    char suit[8];
    switch(c->suit) {
        case SPADE:
            sprintf(suit, "Spade");
            break;
        case CLUB:
            sprintf(suit, "Club");
            break;
        case HEART:
            sprintf(suit, "Heart");
            break;
        case DIAMOND:
            sprintf(suit, "Diamond");
            break;
        default:
            sprintf(suit, "ERROR");
            break;
    }
    char face[6];
    switch(c->value) {
        case ACE:
            sprintf(face,"Ace");
            break;
        case JACK:
            sprintf(face, "Jack");
            break;
        case QUEEN:
            sprintf(face, "Queen");
            break;
        case KING:
            sprintf(face, "King");
            break;
        default:
            sprintf(face, "%d", c->value);
            break;
    }
    sprintf(buffer, "%s of %s", face, suit);
}

Card* makeCard(FaceValue val, Suit s) {
    Card *c = (Card*)malloc(sizeof(Card));
    if (!c) {
        return NULL;
    }
    c->value = val;
    c->suit = s;
    c->toString = CARD_to_string;
    return c;
}

Deck makeDeck(void) {
    Deck cards = (Deck) calloc(NUM_CARDS, sizeof(Card));
    if (!cards) {
        return NULL;
    }
    int i;
    for (i = 0; i < NUM_CARDS; i++) {
        cards[i] = makeCard((FaceValue)(i%13 + 1), (Suit)(i/13));
    }
    return cards;
}

void freeCard(Card* card) {
    free(card);
}

void freeDeck(Deck deck) {
    for (int i = 0; i < NUM_CARDS; i++) {
        freeCard(deck[i]);
    }
    free(deck);
}

void swap(Deck v, int i, int j) {
    Card *temp;

    temp = v[i];
    v[i] = v[j];
    v[j] = temp;
}

void randomize(Deck deck, int numRandoms) {
    int i = 0;
    int a = 0;
    int b = 0;
    for (i = 0; i < numRandoms; i++) {
        a = i % NUM_CARDS;
        do {
            b = rand() % NUM_CARDS;
        } while (a == b);
        swap(deck, a, b);
    }
}

File: cards.swift
import Foundation

public struct Card: CustomStringConvertible {
    public enum Suit: Int {
        case Spade = 0
        case Club
        case Heart
        case Diamond
    }
    public enum FaceValue: Int {
        case Ace = 1
        case Two = 2
        case Three = 3
        case Four = 4
        case Five = 5
        case Six = 6
        case Seven = 7
        case Eight = 8
        case Nine = 9
        case Ten = 10
        case Jack = 11
        case Queen = 12
        case King = 13
    }

    private(set) public var suit: Suit
    private(set) public var value: FaceValue

    public var description: String {
        return "\(value) of \(suit)"
    }
}
public struct Deck: CustomStringConvertible {
    private var deck: [Card]

    public init() {
        deck = [Card]()
        for s in 0..<4 {
            for v in 1...13 {
                deck.append(Card(suit: Card.Suit(rawValue: s)!, value: Card.FaceValue(rawValue: v)!))
            }
        }
    }

    public mutating func shuffle() {
        deck.sort { (_, _) -> Bool in
            return arc4random() % 2 == 0
        }
    }

    public var description: String {
        var deckString = ""
        for (i, card) in deck.enumerated() {
            deckString.append(card.description)
            if (i % 4 == 3) {
                deckString.append("\n")
            } else {
                deckString.append("\t")
            }
        }
        return deckString
    }
}